feat: Send email when charge table has finished syncing

rafaeelaudibert opened the pull request:
When a Stripe `charge` table finishes syncing, we wanna send an email to the members of the team to let them know the sync has finished and `Revenue Analytics` is ready to be used

This is WIP, we havent added the flag yet to avoid sending this email over and over again, wanna collect feedback from DW team

rafaeelaudibert commented on posthog/temporal/data_imports/pipelines/pipeline/pipeline.py:
```diff
@@ -270,6 +271,9 @@ def _post_run_operations(self, row_count: int):
         self._logger.debug("Updating last synced at timestamp on schema")
         _update_last_synced_at_sync(self._schema, self._job)
 
+        self._logger.debug("Notifying revenue analytics that sync has completed")
+        _notify_revenue_analytics_that_sync_has_completed(self._schema)
```
@PostHog/team-data-warehouse is this approach even reasonable, or do you suggest something else to detect that we've finished syncing a source? I'd love us to trigger *something* when the whole source finishes syncing the first time rather than the schema, but I'm not sure that's possible.

EDsCODE replied:
Schemas are decoupled from sources so that they can be independently run so there's no *easy* way to wait for the entire source (because a schema can be set to run multiple times while another schema might be once in a while).

rafaeelaudibert replied:
@EDsCODE 

> Schemas are decoupled from sources so that they can be independently run so there's no easy way to wait for the entire source (because a schema can be set to run multiple times while another schema might be once in a while).

> Is the goal here to have an initial email notification or is notifying everytime the goal?

Ok, so is it acceptable to include a flag in the schema to be able to send this notification only once? It sounds slightly egregious, but that might be all right? Maybe I can add it as a `metadata` JSONB column that holds any crap we might need?

> I would move this logic to update_external_data_job_model activity. Or even a separate activity. There's still important data operations happening after this logic so if things fail, it wouldn't make sense that an email job has been queued. This would make sense as a last activity in the workflow and eventually extendable into more notification logic if we choose to build it out more

Yeah, happy to do so, and to make it non-retriable as well, this is veeery low-priority

Gilbert09 replied:
> Ok, so is it acceptable to include a flag in the schema to be able to send this notification only once? It sounds slightly egregious, but that might be all right? Maybe I can add it as a metadata JSONB column that holds any crap we might need?

I'd rather this was decoupled from data warehouse / schemas - may be better placed in a revenue analytics table with team/project ID instead? The less clutter in warehouse database tables the better 🙏

Gilbert09 replied:
> I would move this logic to update_external_data_job_model activity. Or even a separate activity.

Either this, or wrap your stuff in a try/catch that just passes - starting up another activity is okay, but we do pay for the number of activities we run.

rafaeelaudibert replied:
@Gilbert09 

> I'd rather this was decoupled from data warehouse / schemas - may be better placed in a revenue analytics table with team/project ID instead? The less clutter in warehouse database tables the better 🙏

Ok, that's reasonable, I need a table for base currency too, so that sounds good to me, will rework this to have a separate table

> Either this, or wrap your stuff in a try/catch that just passes - starting up another activity is okay, but we do pay for the number of activities we run.

Ok, this activity would run very often, will just make this failproof - to the best of Python's possibilities

rafaeelaudibert commented on posthog/tasks/email.py:
```diff
@@ -425,6 +426,26 @@ def send_two_factor_auth_backup_code_used_email(user_id: int) -> None:
+        },
+    )
+
+    for membership in get_members_to_notify(team):
+        message.add_recipient(email=membership.user.email, name=membership.user.first_name)
+    message.send(send_async=False)
```
Maybe I should just use `send_message_to_all_staff_users` instead? Probably easier than adding all of these permissions, right?

rafaeelaudibert commented on posthog/temporal/data_imports/pipelines/pipeline_sync.py:
Please ignore, can remove these changes, this was just me testing and learning about the code

rafaeelaudibert commented on posthog/tasks/email.py:
```diff
@@ -425,6 +426,26 @@ def send_two_factor_auth_backup_code_used_email(user_id: int) -> None:
+    team: Team = Team.objects.get(pk=team_id)
+    source: ExternalDataSource = ExternalDataSource.objects.get(team=team, pk=source_id)
+
+    message = EmailMessage(
+        use_http=True,
+        campaign_key=f"revenue_analytics_sync_complete_{team_id}_{source_id}",
+        template_name="revenue_analytics_sync_complete",
+        subject=f"Revenue analytics sync complete for {source.source_type}",
+        template_context={
+            "source_type": source.source_type,
+        },
+    )
```
Template is not live yet, need to get this sorted in customer.io

EDsCODE reviewed (commented):
Is the goal here to have an initial email notification or is notifying everytime the goal?

Gilbert09 commented on posthog/temporal/data_imports/pipelines/pipeline/utils.py:
```diff
@@ -398,6 +400,20 @@ def _update_last_synced_at_sync(schema: ExternalDataSchema, job: ExternalDataJob
+    if (
+        schema.source.source_type == ExternalDataSource.Type.STRIPE
+        and schema.source.revenue_analytics_enabled
+        and schema.source.created_at < datetime.now() - datetime.timedelta(minutes=5)
```
Guessing this is checking for new sources? If so, this will often be False depending on when a user connects the source (e.g. time of day) and how big their stripe table is - paging Stripe takes _a long time_

rafaeelaudibert replied:
I wanna avoid sending emails if the sync is like really fast, because they'll probably be using the app still, and I don't wanna send them an email if it took like 2min. This is checking whether it took more than 5 minutes to run the sync

Gilbert09 commented on posthog/temporal/data_imports/pipelines/pipeline/utils.py:
```diff
@@ -398,6 +400,20 @@ def _update_last_synced_at_sync(schema: ExternalDataSchema, job: ExternalDataJob
+        schema.source.source_type == ExternalDataSource.Type.STRIPE
+        and schema.source.revenue_analytics_enabled
+        and schema.source.created_at < datetime.now() - datetime.timedelta(minutes=5)
+        and schema.name == STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER
```
Just checked out this `STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER` value - it may be better placed somewhere like `posthog/temporal/data_imports/pipelines/stripe/settings.py` - e.g. close to the actual source so that of the schema name does change, it'll be obvious that this would need updating too, but this is very much a nit, so feel free to ignore

rafaeelaudibert replied:
Ok, great call, will have a PR making that better

rafaeelaudibert committed:
feat: Introduce new DB field to detect revenue analytics readiness

This field is used to indicate whether we should let customers know about Revenue Analytics or not, it'll be used in a follow-up commit

rafaeelaudibert committed:
feat: Send email on first revenue-related sync

If we detect we've finished syncing charges AND we havent ever notified the user about that, we'll send them an email letting them know Revenue Analytics is ready to be used.

Email is still pending in customer.io

rafaeelaudibert committed:
refacotr: Extract Stripe resource names to constants (#31744)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Gilbert09 commented on posthog/temporal/data_imports/pipelines/pipeline/pipeline.py:
```diff
@@ -270,6 +271,9 @@ def _post_run_operations(self, row_count: int):
         _update_last_synced_at_sync(self._schema, self._job)
 
+        self._logger.debug("Notifying revenue analytics that sync has completed")
+        _notify_revenue_analytics_that_sync_has_completed(self._schema)
```
+1, the `logger` param is missing here

Gilbert09 commented on posthog/temporal/data_imports/pipelines/pipeline/utils.py:
```diff
@@ -398,6 +401,29 @@ def _update_last_synced_at_sync(schema: ExternalDataSchema, job: ExternalDataJob
+    except Exception as e:
+        # Silently fail, we don't want this to crash the pipeline
+        # Sending an email is not critical to the pipeline
+        logger.exception(f"Error notifying revenue analytics that sync has completed: {e}")
```
You may wanna use `capture_exception(e)` here too so that these errors come through to the error tracking product

Gilbert09 commented on posthog/temporal/data_imports/pipelines/stripe/constants.py:
```diff
@@ -0,0 +1,8 @@
+ACCOUNT_RESOURCE_NAME = "Account"
+BALANCE_TRANSACTION_RESOURCE_NAME = "BalanceTransaction"
+CHARGE_RESOURCE_NAME = "Charge"
+CUSTOMER_RESOURCE_NAME = "Customer"
+INVOICE_RESOURCE_NAME = "Invoice"
+PRICE_RESOURCE_NAME = "Price"
+PRODUCT_RESOURCE_NAME = "Product"
+SUBSCRIPTION_RESOURCE_NAME = "Subscription"
```
thanks for this change - good improvement

rafaeelaudibert replied:
BTW I tried changing this in a separate PR for easier eview, but I ended up creating conflicting changes 😞

Gilbert09 reviewed (approved):
Looks good!

rafaeelaudibert committed:
refactor: Use PH campaign rather than send email

Let's dogfood our product with a low-stakes email

rafaeelaudibert committed:
refactor: format

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

rafaeelaudibert committed:
fix: fix types

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

rafaeelaudibert committed:
fix: Fix tests by refreshing from DB

rafaeelaudibert committed:
fix: Include reverse_code for migration

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rafaeelaudibert committed:
fix: Format Python

rafaeelaudibert committed:
fix: Fix wrong import

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

rafaeelaudibert committed:
fix: Import from correct place

rafaeelaudibert committed:
fix: Fix billing snapshot flapping all the time

rafaeelaudibert committed:
feat: Sync tree view locally

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

rafaeelaudibert committed:
fix: Properly enable revenue analytics on tests
